src/meta/typeutils.go

```diff
@@ -120,204 +120,205 @@ func wrap(typ byte, byteFields []uint8, args ...string) string {
+func (t Type) UnwrapInstancePropertyFetch() (typ Type, propName string) {
+	rawType, propName := t.unwrap2()
+	typ = NewType(rawType)
+	return typ, propName
 }
 
-func WrapFunctionCall(funcName string) string {
-	return wrap(WFunctionCall, nil, funcName)
+func WrapFunctionCall(funcName string) Type {
+	return NewType(wrap(WFunctionCall, nil, funcName))
 }
 
-func UnwrapFunctionCall(s string) (funcName string) {
-	return unwrap1(s)
+func (t Type) UnwrapFunctionCall() (funcName string) {
+	return t.unwrap1()
 }
 
-func WrapArrayOf(typ string) string {
-	return wrap(WArrayOf, nil, typ)
+func WrapArrayOf(typ Type) Type {
+	return NewType(wrap(WArrayOf, nil, typ.String()))
 }
 
-func UnwrapArrayOf(s string) (typ string) {
-	return unwrap1(s)
+func (t Type) UnwrapArrayOf() (typ Type) {
+	return NewType(t.unwrap1())
 }
 
-func WrapArray2(ktyp, vtyp string) string {
-	// TODO: actually support types of keys
-	return WrapArrayOf(vtyp)
-}
-
-func WrapElemOf(typ string) string {
+func WrapElemOf(typ Type) Type {
 	// ElemOf(ArrayOf(typ)) == typ
 	if len(typ) >= 1+stringLenBytes && typ[0] == WArrayOf {
 		return typ[1+stringLenBytes:]
 	}
 
-	return wrap(WElemOf, nil, typ)
+	return NewType(wrap(WElemOf, nil, typ.String()))
 }
 
-func UnwrapElemOf(s string) (typ string) {
-	return unwrap1(s)
+func (t Type) UnwrapElemOf() (typ Type) {
+	return NewType(t.unwrap1())
 }
 
-func WrapElemOfKey(typ, key string) string {
+func WrapElemOfKey(typ Type, key string) Type {
 	if len(typ) >= 1+stringLenBytes && typ[0] == WArrayOf {
 		return typ[1+stringLenBytes:]
 	}
-	return wrap(WElemOfKey, nil, typ, key)
+	return NewType(wrap(WElemOfKey, nil, typ.String(), key))
 }
 
-func UnwrapElemOfKey(s string) (typ, key string) {
-	return unwrap2(s)
+func (t Type) UnwrapElemOfKey() (typ Type, key string) {
+	rawType, key := t.unwrap2()
+	typ = NewType(rawType)
+	return typ, key
 }
 
-func WrapGlobal(varName string) string {
-	return wrap(WGlobal, nil, varName)
+func WrapGlobal(varName string) Type {
+	return NewType(wrap(WGlobal, nil, varName))
 }
 
-func UnwrapGlobal(s string) (varName string) {
-	return unwrap1(s)
+func (t Type) UnwrapGlobal() (varName string) {
+	return t.unwrap1()
 }
 
-func WrapConstant(constName string) string {
-	return wrap(WConstant, nil, constName)
+func WrapConstant(constName string) Type {
+	return NewType(wrap(WConstant, nil, constName))
 }
 
-func UnwrapConstant(s string) (constName string) {
-	return unwrap1(s)
+func (t Type) UnwrapConstant() (constName string) {
+	return t.unwrap1()
 }
 
-func formatType(s string) (res string) {
-	if len(s) == 0 || s[0] >= WMax {
-		return s
+func (t Type) Format() (res string) {
+	if t.IsEmpty() || !t.IsLazy() {
+		return t.String()
 	}
 
 	defer func() {
 		if r := recover(); r != nil {
-			res = fmt.Sprintf("panic!(orig='%s', hex='%x')", s, s)
+			res = fmt.Sprintf("panic!(orig='%s', hex='%x')", t, t)
 		}
 	}()
 
-	switch s[0] {
+	switch t[0] {
 	case WGlobal:
-		return "global_$" + formatType(UnwrapGlobal(s))
```

Comment:
Why did you remove formatting here? It is not present in the new `UnwrapGlobal` method

Reply:
Now there is, thanks.